server/src/main/java/org/candlepin/audit/EventSinkImpl.java

```diff
@@ -187,13 +135,15 @@ public void queueEvent(Event event) {
         log.debug("Queuing event: {}", event);
 
         try {
-            ClientSession session = getClientSession();
+            // Lazily initialize the session and producer when the first
+            // message gets queued.
+            initializeMessageSending();
```

Comment:
The intent of doing it this way is to ensure that we do not create a new connection to the broker if it isn't required (expensive) -- only when a message is queued , not every request requires message sending.

The session is required for queuing, sending and rollback, so unless I'm missing something, the only way to do this would be to encapsulate session/producer in a new class (another level of indirection), which I'm not sure we benefit much from since this is isolated to a single class already. 

initializeMessageSending is only being used in two places and getQueueInfo shouldn't be in this class to begin with IMO.

Thoughts?

Reply:
I think we're both have the same idea, just with different ways of implementing it.

All I'm saying here is that rather than use the instance variables directly, since we're calling some kind of method at the beginning of all of our queuing/sending bits, it makes more sense to me to do something similar to the original code where we call getSession or some such, which itself does the work of making sure we only have one session/instance and then initializes it as necessary.

This gives us the advantage of keeping all of the setup in one place without encouraging poking directly at the "shared" instance variables, and more gracefully prepares us for any subclassing we may or may not want to do in the future.